core/iwasm/libraries/lib-wasi-threads/test/spawn_stress_test.c

```diff
@@ -0,0 +1,161 @@
+#include "sync_primitives.h"
+#else
+#include <pthread.h>
+#endif
+
+#include "wasi_thread_start.h"
+
+enum CONSTANTS {
+    NUM_ITER = 1000000,
+    NUM_RETRY = 5,
+    SECOND = 1000 * 1000 * 1000, /* 1 second */
+    TIMEOUT = 50LL * SECOND,
+    MAX_NUM_THREADS = 4,
+};
+
+typedef struct {
+    start_args_t base;
+    int th_done;
+    int thread_id;
+    int value;
+} shared_t;
+
+unsigned int factorised_number = 1;
+unsigned int prime_numbers = 0;
+double percentage = 0.1;
+
+// Inited threads - Deinited threads. Should always be zero in the end. Otherwise it's a leak.
+int thread_init_delta = 0;
+
+pthread_mutex_t mutex;
+shared_t threads[MAX_NUM_THREADS];
+
+bool is_prime(unsigned int num) {
+    for (unsigned int  i = 2; i <= (unsigned int)(sqrt(num)); ++i) {
+        if (num % i == 0) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+void
+__wasi_thread_start_C(int thread_id, int *start_arg)
+{
+    shared_t *data = (shared_t *)start_arg;
+
+    if (is_prime(data->value)) {
+        __atomic_fetch_add(&prime_numbers, 1, __ATOMIC_SEQ_CST);
+    }
+
+    __atomic_store_n(&data->th_done, 1, __ATOMIC_SEQ_CST);
+    __builtin_wasm_memory_atomic_notify(&data->th_done, 1);
+}
+
+unsigned int validate() {
+    unsigned int counter = 0;
+    for (unsigned int i = 2; i <= factorised_number; ++i) {
+        counter += is_prime(i);
+    }
+
+    return counter;
+}
+
+int
+main(int argc, char **argv)
+{
+    pthread_mutex_init(&mutex, NULL);
+
+    for (int i = 0; i < MAX_NUM_THREADS; ++i) {
+        pthread_mutex_lock(&mutex);
+
+        // Init a new thread
+        {
+            assert(start_args_init(&threads[i].base));
+            factorised_number++;
+            threads[i].value = factorised_number;
+            thread_init_delta++;
+        }
+        
+        threads[i].thread_id = __wasi_thread_spawn(&threads[i]);
+        assert(threads[i].thread_id > 0 && "Thread creation should succeed");
+        pthread_mutex_unlock(&mutex);
+    }
+
+    while (!pthread_mutex_lock(&mutex) && factorised_number < NUM_ITER) {
+        if (factorised_number > (unsigned int)(NUM_ITER * percentage)) {
+            printf("Stress test %d %% finished\n", (unsigned int)(percentage * 100));
+            percentage += 0.1;
+        }
+        
+        // Locking the mutex and checking whether every thread is done as we received the notify
+        for (int j = 0; j < MAX_NUM_THREADS && factorised_number < NUM_ITER; ++j) {
+            // fprintf(stderr, "Waiting for thread %d \n", j);
+            pthread_mutex_unlock(&mutex);
+            assert(__builtin_wasm_memory_atomic_wait32(&threads[j].th_done, 0, TIMEOUT) != 2);
+                // Deinit the expired thread
+            {
+                start_args_deinit(&threads[j].base);
+                thread_init_delta--;
+            }
+            
+            // Invalidated the expired thread
+            {
+                threads[j].th_done = 0;
```

Comment:
consider using `memset(&threads[j], 0, sizeof(threads[j]))` so if we ever add any new fields to the structure, they'll automatically be zero-ed. thread ID 0 is invalid too, no need to set it to -1.